### PR TITLE
[wip] improved wasm API for JavaScript.

### PR DESCRIPTION
**Pull request by burdettadam**
This could be the start of something magical!
my first attempts at creating a cred_def in javascript. 
Signed-off-by: Adam Burdett <burdettadam@gmail.com>

### CODE REVIEW

File: libursa/src/wasm/bls.rs [hunk 1]
```diff
@@ -22,24 +22,27 @@ impl Generator {
...
-        Ok(SignKey(maperr!(bls::SignKey::new(None))))
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Result<JsValue, JsValue>  {
+        Ok(JsValue::from_serde(&SignKey(bls::SignKey::new(None)?)).unwrap())
     }
-
-    pub fn from_seed(seed: &[u8]) -> Result<SignKey, JsValue> {
-        Ok(SignKey(maperr!(bls::SignKey::new(Some(seed)))))
+    pub fn from_seed(seed: JsValue) -> Result<JsValue, JsValue> {
+        let _seed: Vec<u8> = seed.into_serde().unwrap();
```
Comment (mac-arrap) [inline libursa/src/wasm/bls.rs]: Please unwrap this correctly and don't use the unwrap function in production code

File: libursa/src/wasm/cl.rs [hunk 2]
```diff
@@ -116,14 +130,36 @@ pub struct CredentialDefinition {
...
 #[wasm_bindgen]
 impl MasterSecret {
-    pub fn new() -> Result<MasterSecret, JsValue> {
-        Ok(MasterSecret(maperr!(
-            cl::prover::Prover::new_master_secret()
-        )))
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Result<JsValue, JsValue>{
+        Ok(JsValue::from_serde(&MasterSecret(
+            cl::prover::Prover::new_master_secret()?
```
Comment (mac-arrap) [inline libursa/src/wasm/cl.rs]: and here

File: libursa/src/wasm/cl.rs [hunk 3]
```diff
@@ -116,14 +130,36 @@ pub struct CredentialDefinition {
...
+        let attrs: JsCredentialSchema = js_attrs.into_serde().unwrap();
+        let hidden_attrs: JsNonCredentialSchema = js_hidden_attrs.into_serde().unwrap();
+
+        let (credential_public_key, credential_private_key, credential_key_correctness_proof) =
+        cl::issuer::Issuer::new_credential_def(&attrs.0, &hidden_attrs.0, support_revocation)?;
+        Ok(JsValue::from_serde(&CredentialDefinition {
+            publicKey: CredentialPublicKey(credential_public_key),
+            privateKey: CredentialPrivateKey(credential_private_key),
+            keyCorrectnessProof: CredentialKeyCorrectnessProof(credential_key_correctness_proof),
+        }).unwrap())
```
Comment (mac-arrap) [inline libursa/src/wasm/cl.rs]: here

File: libursa/src/wasm/bls.rs [hunk 4]
```diff
@@ -22,24 +22,27 @@ impl Generator {
...
-        Ok(SignKey(maperr!(bls::SignKey::from_bytes(bytes))))
+    pub fn from_bytes(bytes: JsValue) -> Result<JsValue, JsValue> {
+        let _bytes: Vec<u8> = bytes.into_serde().unwrap();
+        Ok(JsValue::from_serde(&SignKey(bls::SignKey::from_bytes(&_bytes)?)).unwrap())
     }
 
-    pub fn to_bytes(&self) -> Result<Vec<u8>, JsValue> {
-        Ok(self.0.as_bytes().to_vec())
+    pub fn to_bytes(&self) -> Result<JsValue, JsValue> {
+        Ok(JsValue::from_serde(&self.0.as_bytes().to_vec()).unwrap())
```
Comment (mac-arrap) [inline libursa/src/wasm/bls.rs]: here

File: libursa/src/wasm/bls.rs [hunk 5]
```diff
@@ -22,24 +22,27 @@ impl Generator {
...
-        Ok(SignKey(maperr!(bls::SignKey::new(Some(seed)))))
+    pub fn from_seed(seed: JsValue) -> Result<JsValue, JsValue> {
+        let _seed: Vec<u8> = seed.into_serde().unwrap();
+        Ok(JsValue::from_serde(&SignKey(bls::SignKey::new(Some(&_seed))?)).unwrap())
     }
 
-    pub fn from_bytes(bytes: &[u8]) -> Result<SignKey, JsValue> {
-        Ok(SignKey(maperr!(bls::SignKey::from_bytes(bytes))))
+    pub fn from_bytes(bytes: JsValue) -> Result<JsValue, JsValue> {
+        let _bytes: Vec<u8> = bytes.into_serde().unwrap();
```
Comment (mac-arrap) [inline libursa/src/wasm/bls.rs]: here

File: libursa/src/wasm/bls.rs [hunk 6]
```diff
@@ -22,24 +22,27 @@ impl Generator {
...
+#[derive(Serialize, Deserialize)]
 pub struct SignKey(bls::SignKey);
 
 #[wasm_bindgen]
 impl SignKey {
-    pub fn new() -> Result<SignKey, JsValue> {
-        Ok(SignKey(maperr!(bls::SignKey::new(None))))
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Result<JsValue, JsValue>  {
+        Ok(JsValue::from_serde(&SignKey(bls::SignKey::new(None)?)).unwrap())
```
Comment (mac-arrap) [inline libursa/src/wasm/bls.rs]: here

File: libursa/src/wasm/bls.rs (hunk 1 above)
Comment (burdettadam) [reply libursa/src/wasm/bls.rs]: I updated the code to use map_err instead of unwrap.

File: libursa/src/wasm/cl.rs [hunk 7]
```diff
@@ -23,10 +28,15 @@ impl CredentialSchema {
...
+#[derive(Serialize, Deserialize)]
+pub struct JsNonCredentialSchema(cl::NonCredentialSchema);
+
+#[wasm_bindgen]
+#[derive(Serialize, Deserialize)]
 pub struct NonCredentialSchema(cl::NonCredentialSchemaBuilder);
 
 #[wasm_bindgen]
-impl NonCredentialSchema {
+impl NonCredentialSchema{
```
Comment (mac-arrap) [inline libursa/src/wasm/cl.rs]: please add space here to keep coding standards

File: libursa/src/wasm/cl.rs [hunk 8]
```diff
@@ -116,14 +130,36 @@ pub struct CredentialDefinition {
 }
 
 #[wasm_bindgen]
+impl CredentialDefinition {
```
Comment (mac-arrap) [inline libursa/src/wasm/cl.rs]: @mikelodder7 Do you approve of having cred_defs in ursa? Please review

**Pull request closed without merging**